desc/integrals/_interp_utils.py

```diff
@@ -0,0 +1,754 @@
+#  spectral methods section 19.6. Can isolate interval to search for root by
+#  observing whether B - 1/pitch changes sign at extrema. Only need to do
+#  evaluate Chebyshev series at quadrature points once, and can use that to
+#  compute the integral for every pitch. The integral will converge rapidly
+#  since a low order polynomial approximates |B| well in between adjacent
+#  extrema. This is cheaper and non-iterative, so jax and gpu will like it.
+#  Implementing 1 and 2 will remove all eigenvalue solves from computation.
+#  2 is a larger improvement than 1. Implement this in later PR.
+chebroots_vec = jnp.vectorize(chebroots, signature="(m)->(n)")
+
+
+def cheb_pts(n, domain=(-1, 1), lobatto=False):
+    """Get ``n`` Chebyshev points mapped to given domain.
+
+    Warnings
+    --------
+    This is a common definition of the Chebyshev points (see Boyd, Chebyshev and
+    Fourier Spectral Methods p. 498). These are the points demanded by discrete
+    cosine transformations to interpolate Chebyshev series because the cosine
+    basis for the DCT is defined on [0, π]. They differ in ordering from the
+    points returned by ``numpy.polynomial.chebyshev.chebpts1`` and
+    ``numpy.polynomial.chebyshev.chebpts2``.
+
+    Parameters
+    ----------
+    n : int
+        Number of points.
+    domain : tuple[float]
+        Domain for points.
+    lobatto : bool
+        Whether to return the Gauss-Lobatto (extrema-plus-endpoint)
+        instead of the interior roots for Chebyshev points.
+
+    Returns
+    -------
+    pts : jnp.ndarray
+        Shape (n, ).
+        Chebyshev points mapped to given domain.
+
+    """
+    N = jnp.arange(n)
+    if lobatto:
+        y = jnp.cos(jnp.pi * N / (n - 1))
+    else:
+        y = jnp.cos(jnp.pi * (2 * N + 1) / (2 * n))
+    return bijection_from_disc(y, domain[0], domain[-1])
+
+
+def fourier_pts(n):
+    """Get ``n`` Fourier points in [0, 2π)."""
+    # [0, 2π] instead of [-π, π] required to match our definition of α.
+    return 2 * jnp.pi * jnp.arange(n) / n
+
+
+def harmonic(a, n, axis=-1):
+    """Spectral coefficients of the Nyquist trigonometric interpolant.
+
+    Parameters
+    ----------
+    a : jnp.ndarray
+        Fourier coefficients ``a=rfft(f,norm="forward",axis=axis)``.
+    n : int
+        Spectral resolution of ``a``.
+    axis : int
+        Axis along which coefficients are stored.
+
+    Returns
+    -------
+    h : jnp.ndarray
+        Nyquist trigonometric interpolant coefficients.
+        Coefficients ordered along ``axis`` of size ``n`` to match ordering of
+        [1, cos(x), ..., cos(nx), sin(x), sin(2x), ..., sin(nx)] basis.
+
+    """
+    is_even = (n % 2) == 0
+    # cos(nx) coefficients
+    an = 2.0 * (
+        a.real.at[Index.get(0, axis, a.ndim)]
+        .divide(2.0)
+        .at[Index.get(-1, axis, a.ndim)]
+        .divide(1.0 + is_even)
+    )
+    # sin(nx) coefficients
+    bn = -2.0 * take(
```

Comment:
why is the sin coef negative here?

Reply:
maybe the best way to explain this is to look at the evaluation in `interp_rfft`. In particular
```python
(basis * a).real.sum(axis=-1)
```
where both `basis` and `a` are complex.
Taking the real part is equivalent to computing cos(n xq) - sin(n xq).